### PR TITLE
[serve] Fix state when a bad runtime env is used for a deployment

### LINKED ISSUE #38407: [Serve] Runtime env failure causing serve unavailable.
### What happened + What you expected to happen

script:

```
from ray import serve

@serve.deployment(ray_actor_options={"runtime_env": {"pip": ["Not_exist"]}})
class Model:
    def __call__(self):
        return 123
entry = Model.bind()
```

serve run file_name:entry

Seeing 
```
(ServeController pid=72076) Traceback (most recent call last):
(ServeController pid=72076)   File "/Users/sihanwang/code/ray/python/ray/serve/controller.py", line 351, in run_control_loop
(ServeController pid=72076)     any_recovering = self.deployment_state_manager.update()
(ServeController pid=72076)   File "/Users/sihanwang/code/ray/python/ray/serve/_private/deployment_state.py", line 2644, in update
(ServeController pid=72076)     self._deployment_scheduler.on_deployment_deleted(tag)
(ServeController pid=72076)   File "/Users/sihanwang/code/ray/python/ray/serve/_private/deployment_scheduler.py", line 111, in on_deployment_deleted
(ServeController pid=72076)     assert not self._launching_replicas[deployment_name]
(ServeController pid=72076) AssertionError
```

After this happens, serve instance is not able to deploy anything.

### Versions / Dependencies

master

### Reproduction script

provided

### Issue Severity

None

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

See issue.

## Related issue number

Closes https://github.com/ray-project/ray/issues/38407

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/deployment_state.py [hunk 1]
```diff
@@ -571,6 +571,19 @@ def check_ready(self) -> Tuple[ReplicaStartupStatus, Optional[str]]:
                     "the replica will be stopped."
                 )
                 return ReplicaStartupStatus.FAILED, str(e.as_instanceof_cause())
+            except RuntimeEnvSetupError as e:
+                msg = (
+                    f"Exception when allocating replica '{self._replica_tag}': {str(e)}"
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/deployment_state.py:576]: Should we use `str(e)` or `traceback.format_exc()` here?

File: python/ray/serve/tests/test_deploy_2.py [hunk 2]
```diff
@@ -427,5 +427,25 @@ def check_health(self):
         time.sleep(0.1)
 
 
+def test_deploy_bad_working_dir_deployment(serve_instance):
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_deploy_2.py:430]: There's no `working_dir` in this test.

```suggestion
def test_deploy_bad_runtime_env_deployment(serve_instance):
```

Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_deploy_2.py:430]: Do we already have a test for a bad app-level `runtime_env`?

File: python/ray/serve/_private/deployment_state.py (hunk 1 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:576]: As far as I can tell from the examples I ran, they are the same except `traceback.format_exc()` has an extra block that isn't really useful for the user (like what happens when you do `raise from`). Also, `RuntimeEnvSetupError` has its own custom `__str__` method implemented - that's why I decided to go with `str(e)`.

File: python/ray/serve/tests/test_deploy_2.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/tests/test_deploy_2.py:430]: We have one in `test_cli`!

Comment (zcin) [reply python/ray/serve/tests/test_deploy_2.py]: Whoops my bad, renamed!

File: python/ray/serve/_private/deployment_state.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/deployment_state.py:576]: Got it, that makes sense. Thanks for the detailed explanation!

**Review by shrekris-anyscale — APPROVED** [review]
Nice work!

**Review by sihanwang41 — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Pull request merged**